File: urbanairship-react-native/android/src/main/java/com/urbanairship/reactnative/ReactAutopilot.java
/* Copyright Urban Airship and Contributors */

package com.urbanairship.reactnative;

import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.preference.PreferenceManager;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.XmlRes;

import com.facebook.react.views.view.ColorUtil;
import com.urbanairship.AirshipConfigOptions;
import com.urbanairship.Autopilot;
import com.urbanairship.Logger;
import com.urbanairship.PrivacyManager;
import com.urbanairship.UAirship;
import com.urbanairship.actions.DeepLinkListener;
import com.urbanairship.analytics.Analytics;
import com.urbanairship.channel.AirshipChannelListener;
import com.urbanairship.json.JsonList;
import com.urbanairship.json.JsonMap;
import com.urbanairship.json.JsonValue;
import com.urbanairship.messagecenter.InboxListener;
import com.urbanairship.messagecenter.MessageCenter;
import com.urbanairship.push.NotificationActionButtonInfo;
import com.urbanairship.push.NotificationInfo;
import com.urbanairship.push.NotificationListener;
import com.urbanairship.push.PushListener;
import com.urbanairship.push.PushMessage;
import com.urbanairship.reactnative.events.DeepLinkEvent;
import com.urbanairship.reactnative.events.InboxUpdatedEvent;
import com.urbanairship.reactnative.events.NotificationResponseEvent;
import com.urbanairship.reactnative.events.PushReceivedEvent;
import com.urbanairship.reactnative.events.RegistrationEvent;
import com.urbanairship.reactnative.events.ShowInboxEvent;

import okhttp3.internal.Util;

/**
 * Module's autopilot to customize Urban Airship.
 */
public class ReactAutopilot extends Autopilot {
    public static final String EXTENDER_MANIFEST_KEY = "com.urbanairship.reactnative.AIRSHIP_EXTENDER";

    private AirshipConfigOptions configOptions;

    @Override
    public void onAirshipReady(@NonNull UAirship airship) {
        super.onAirshipReady(airship);

        PluginLogger.setLogLevel(airship.getAirshipConfigOptions().logLevel);

        PluginLogger.debug("Airship React Native version: %s, SDK version: %s", BuildConfig.MODULE_VERSION, UAirship.getVersion());

        final Context context = UAirship.getApplicationContext();

        airship.setDeepLinkListener(new DeepLinkListener() {
            @Override
            public boolean onDeepLink(@NonNull String deepLink) {
                Event event = new DeepLinkEvent(deepLink);
                EventEmitter.shared().sendEvent(event);
                return true;
            }
        });

        airship.getPushManager().addPushListener(new PushListener() {
            @Override
            public void onPushReceived(@NonNull PushMessage pushMessage, boolean notificationPosted) {
                if (!notificationPosted) {
                    Event event = new PushReceivedEvent(pushMessage);
                    EventEmitter.shared().sendEvent(event);
                }
            }
        });

        airship.getChannel().addChannelListener(new AirshipChannelListener() {
            @Override
            public void onChannelCreated(@NonNull String channelId) {
                Event event = new RegistrationEvent(channelId, UAirship.shared().getPushManager().getPushToken());
                EventEmitter.shared().sendEvent(event);
            }

            @Override
            public void onChannelUpdated(@NonNull String channelId) {
                Event event = new RegistrationEvent(channelId, UAirship.shared().getPushManager().getPushToken());
                EventEmitter.shared().sendEvent(event);
            }
        });

        airship.getPushManager().setNotificationListener(new NotificationListener() {
            @Override
            public void onNotificationPosted(@NonNull NotificationInfo notificationInfo) {
                Event event = new PushReceivedEvent(notificationInfo);
                EventEmitter.shared().sendEvent(event);
            }

            @Override
            public boolean onNotificationOpened(@NonNull NotificationInfo notificationInfo) {
                Event event = new NotificationResponseEvent(notificationInfo);
                EventEmitter.shared().sendEvent(event);
                return false;
            }

            @Override
            public boolean onNotificationForegroundAction(@NonNull NotificationInfo notificationInfo, @NonNull NotificationActionButtonInfo notificationActionButtonInfo) {
                Event event = new NotificationResponseEvent(notificationInfo, notificationActionButtonInfo);
                EventEmitter.shared().sendEvent(event);
                return false;
            }

            @Override
            public void onNotificationBackgroundAction(@NonNull NotificationInfo notificationInfo, @NonNull NotificationActionButtonInfo notificationActionButtonInfo) {
                Event event = new NotificationResponseEvent(notificationInfo, notificationActionButtonInfo);
                EventEmitter.shared().sendEvent(event);
            }

            @Override
            public void onNotificationDismissed(@NonNull NotificationInfo notificationInfo) {
            }
        });

        // Register a listener for inbox update event
        MessageCenter.shared().getInbox().addListener(new InboxListener() {
            @Override
            public void onInboxUpdated() {
                Event event = new InboxUpdatedEvent(MessageCenter.shared().getInbox().getUnreadCount(), MessageCenter.shared().getInbox().getCount());
                EventEmitter.shared().sendEvent(event);
            }
        });

        ReactAirshipPreferences preferences = ReactAirshipPreferences.shared(context);
        MessageCenter.shared().setOnShowMessageCenterListener(new MessageCenter.OnShowMessageCenterListener() {
            @Override
            public boolean onShowMessageCenter(@Nullable String messageId) {
                if (preferences.isAutoLaunchMessageCenterEnabled()) {
                    return false;
                } else {
                    sendShowInboxEvent(messageId);
                    return true;
                }
            }
        });

        // Set our custom notification provider
        ReactNotificationProvider notificationProvider = new ReactNotificationProvider(context, airship.getAirshipConfigOptions());
        airship.getPushManager().setNotificationProvider(notificationProvider);

        airship.getAnalytics().registerSDKExtension(Analytics.EXTENSION_REACT_NATIVE, BuildConfig.MODULE_VERSION);

        loadCustomNotificationChannels(context, airship);
        loadCustomNotificationButtonGroups(context, airship);

        AirshipExtender extender = createExtender(context);
        if (extender != null) {
            extender.onAirshipReady(context, airship);
        }
    }

    private void loadCustomNotificationChannels(Context context, UAirship airship) {
        String packageName = UAirship.getPackageName();
        @XmlRes int resId = context.getResources().getIdentifier("ua_custom_notification_channels", "xml", packageName);

        if (resId != 0) {
            PluginLogger.debug("Loading custom notification channels");
            airship.getPushManager().getNotificationChannelRegistry().createNotificationChannels(resId);
        }
    }

    private void loadCustomNotificationButtonGroups(Context context, UAirship airship) {
        String packageName = UAirship.getPackageName();
        @XmlRes int resId = context.getResources().getIdentifier("ua_custom_notification_buttons", "xml", packageName);

        if (resId != 0) {
            PluginLogger.debug("Loading custom notification button groups");
            airship.getPushManager().addNotificationActionButtonGroups(context, resId);
        }
    }

    private static void sendShowInboxEvent(@Nullable String messageId) {
        Event event = new ShowInboxEvent(messageId);
        EventEmitter.shared().sendEvent(event);
    }

    @Nullable
    private static AirshipExtender createExtender(@NonNull Context context) {
        ApplicationInfo ai;
        try {
            ai = context.getPackageManager().getApplicationInfo(context.getPackageName(), PackageManager.GET_META_DATA);
            if (ai == null || ai.metaData == null) {
                return null;
            }
        } catch (PackageManager.NameNotFoundException e) {
            return null;
        }

        String classname = ai.metaData.getString(EXTENDER_MANIFEST_KEY);

        if (classname == null) {
            return null;
        }

        try {
            Class<?> extenderClass = Class.forName(classname);
            return (AirshipExtender) extenderClass.newInstance();
        } catch (Exception e) {
            PluginLogger.error(e, "Unable to create extender: " + classname);
        }
        return null;
    }

    @Override
    public boolean isReady(@NonNull Context context) {
        configOptions = loadConfig(context);

        try {
            configOptions.validate();
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    @Nullable
    @Override
    public AirshipConfigOptions createAirshipConfigOptions(@NonNull Context context) {
        return configOptions;
    }

    private AirshipConfigOptions loadConfig(@NonNull Context context) {
        AirshipConfigOptions.Builder builder = AirshipConfigOptions.newBuilder()
                .applyDefaultProperties(context);

        JsonMap config = ReactAirshipPreferences.shared(context).getAirshipConfig();
        if (config == null || config.isEmpty()) {
            return builder.build();
        }

        JsonMap developmentEnvironment = config.opt("development").getMap();
        JsonMap productionEnvironment = config.opt("production").getMap();
        JsonMap defaultEnvironment = config.opt("production").getMap();

        if (developmentEnvironment != null) {
            builder.setDevelopmentAppKey(developmentEnvironment.opt("appKey").getString())
                    .setDevelopmentAppSecret(developmentEnvironment.opt("appSecret").getString());

            String logLevel = developmentEnvironment.opt("logLevel").getString();
            if (logLevel != null) {
                builder.setLogLevel(convertLogLevel(logLevel, Log.DEBUG));
            }
        }

        if (productionEnvironment != null) {
            builder.setProductionAppKey(productionEnvironment.opt("appKey").getString())
                    .setProductionAppSecret(productionEnvironment.opt("appSecret").getString());

            String logLevel = productionEnvironment.opt("logLevel").getString();
            if (logLevel != null) {
                builder.setProductionLogLevel(convertLogLevel(logLevel, Log.ERROR));
            }
        }

        if (defaultEnvironment != null) {
            builder.setAppKey(defaultEnvironment.opt("appKey").getString())
                    .setAppSecret(defaultEnvironment.opt("appSecret").getString());

            String logLevel = defaultEnvironment.opt("logLevel").getString();
            if (logLevel != null) {
                builder.setLogLevel(convertLogLevel(logLevel, Log.ERROR));
            }
        }

        String site = config.opt("site").getString();
        if (site != null) {
            try {
                builder.setSite(parseSite(site));
            } catch (Exception e) {
                PluginLogger.error("Invalid site " + site, e);
            }
        }

        if (config.containsKey("inProduction")) {
            builder.setInProduction(config.opt("inProduction").getBoolean(false));
        }

        if (config.containsKey("isChannelCreationDelayEnabled")) {
            builder.setChannelCreationDelayEnabled(config.opt("isChannelCreationDelayEnabled").getBoolean(false));
        }

        if (config.containsKey("requireInitialRemoteConfigEnabled")) {
            builder.setRequireInitialRemoteConfigEnabled(config.opt("requireInitialRemoteConfigEnabled").getBoolean(false));
        }

        String[] urlAllowList = parseArray(config.opt("urlAllowList"));
        if (urlAllowList != null) {
            builder.setUrlAllowList(urlAllowList);
        }

        String[] urlAllowListScopeJavaScriptInterface = parseArray(config.opt("urlAllowListScopeJavaScriptInterface"));
        if (urlAllowListScopeJavaScriptInterface != null) {
            builder.setUrlAllowListScopeJavaScriptInterface(urlAllowListScopeJavaScriptInterface);
        }

        String[] urlAllowListScopeOpenUrl = parseArray(config.opt("urlAllowListScopeOpenUrl"));
        if (urlAllowListScopeOpenUrl != null) {
            builder.setUrlAllowListScopeOpenUrl(urlAllowListScopeOpenUrl);
        }

        JsonMap chat = config.opt("chat").getMap();
        if (chat != null) {
            builder.setChatSocketUrl(chat.opt("webSocketUrl").optString())
                    .setChatUrl(chat.opt("url").optString());
        }

        JsonMap android = config.opt("android").getMap();
        if (android != null) {
            if (android.containsKey("appStoreUri")) {
                builder.setAppStoreUri(Uri.parse(android.opt("appStoreUri").optString()));
            }

            if (android.containsKey("fcmFirebaseAppName")) {
                builder.setFcmFirebaseAppName(android.opt("fcmFirebaseAppName").optString());
            }

            if (android.containsKey("notificationConfig")) {
                applyNotificationConfig(context, android.opt("notificationConfig").optMap(), builder);
            }
        }

        JsonList enabledFeatures = config.opt("enabledFeatures").getList();
        try {
            if (enabledFeatures != null) {
                builder.setEnabledFeatures(parseFeatures(enabledFeatures));
            }
        } catch (Exception e) {
            PluginLogger.error("Invalid enabled features: " + enabledFeatures);
        }

        return builder.build();
    }

    private void applyNotificationConfig(@NonNull Context context, @NonNull JsonMap notificationConfig, @NonNull AirshipConfigOptions.Builder builder) {
        String icon = notificationConfig.opt(UrbanAirshipReactModule.NOTIFICATION_ICON_KEY).getString();
        if (icon != null) {
            int resourceId = Utils.getNamedResource(context, icon, "drawable");
            builder.setNotificationIcon(resourceId);
        }

        String largeIcon = notificationConfig.opt(UrbanAirshipReactModule.NOTIFICATION_LARGE_ICON_KEY).getString();
        if (largeIcon != null) {
            int resourceId = Utils.getNamedResource(context, largeIcon, "drawable");
            builder.setNotificationLargeIcon(resourceId);
        }

        String accentColor = notificationConfig.opt(UrbanAirshipReactModule.ACCENT_COLOR_KEY).getString();
        if (accentColor != null) {
            builder.setNotificationAccentColor(Utils.getHexColor(accentColor, 0));
        }

        String channelId = notificationConfig.opt(UrbanAirshipReactModule.DEFAULT_CHANNEL_ID_KEY).getString();
        if (channelId != null) {
            builder.setNotificationChannel(channelId);
        }
    }

    private static int convertLogLevel(@NonNull String logLevel, int defaultValue) {
        switch (logLevel) {
            case "verbose":
                return Log.VERBOSE;
            case "debug":
                return Log.DEBUG;
            case "info":
                return Log.INFO;
            case "warning":
                return Log.WARN;
            case "error":
                return Log.ERROR;
            case "none":
                return Log.ASSERT;
        }
        return defaultValue;
    }

    @Nullable
    private static String[] parseArray(@Nullable JsonValue value) {
        if (value == null || !value.isJsonList()) {
            return null;
        }

        String[] result = new String[value.optList().size()];
        for (int i = 0; i <= value.optList().size(); i++) {
            String string = value.optList().get(i).getString();
            if (string == null) {
                PluginLogger.error("Invalid string array: " + value);
                return null;
            }
            result[i] = string;
        }

        return result;
    }

    @PrivacyManager.Feature
    private int parseFeatures(@NonNull JsonList jsonList) {
        int result = PrivacyManager.FEATURE_NONE;
        for (JsonValue value : jsonList) {
            result |= Utils.parseFeature(value.optString());
        }
        return result;
    }

    @AirshipConfigOptions.Site
    @NonNull
    private String parseSite(@NonNull String value) {
        switch (value) {
            case "eu":
                return AirshipConfigOptions.SITE_EU;

            case "us":
                return AirshipConfigOptions.SITE_US;
        }

        throw new IllegalArgumentException("Invalid site: " + value);
    }
}

File: urbanairship-react-native/android/src/main/java/com/urbanairship/reactnative/ReactAutopilot.kt
/* Copyright Urban Airship and Contributors */

package com.urbanairship.reactnative

import android.annotation.SuppressLint
import android.content.Context
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.net.Uri
import android.util.Log
import androidx.annotation.XmlRes
import com.urbanairship.AirshipConfigOptions
import com.urbanairship.Autopilot
import com.urbanairship.Logger
import com.urbanairship.PrivacyManager
import com.urbanairship.UAirship
import com.urbanairship.actions.DeepLinkListener
import com.urbanairship.analytics.Analytics
import com.urbanairship.channel.AirshipChannelListener
import com.urbanairship.json.JsonList
import com.urbanairship.json.JsonMap
import com.urbanairship.json.JsonValue
import com.urbanairship.messagecenter.MessageCenter
import com.urbanairship.preferencecenter.PreferenceCenter
import com.urbanairship.push.NotificationActionButtonInfo
import com.urbanairship.push.NotificationInfo
import com.urbanairship.push.NotificationListener
import com.urbanairship.reactnative.Utils.getHexColor
import com.urbanairship.reactnative.Utils.getNamedResource
import com.urbanairship.reactnative.Utils.parseFeature
import com.urbanairship.reactnative.events.*
import com.urbanairship.reactnative.preferenceCenter.events.DisplayPreferenceCenterEvent

/**
 * Module's autopilot to customize Urban Airship.
 */
class ReactAutopilot : Autopilot() {

    private var configOptions: AirshipConfigOptions? = null

    override fun onAirshipReady(airship: UAirship) {
        super.onAirshipReady(airship)

        PluginLogger.setLogLevel(airship.airshipConfigOptions.logLevel)

        PluginLogger.debug("Airship React Native version: %s, SDK version: %s", BuildConfig.MODULE_VERSION, UAirship.getVersion())

        val context = UAirship.getApplicationContext()

        airship.deepLinkListener = DeepLinkListener { deepLink ->
            val event: Event = DeepLinkEvent(deepLink)
            EventEmitter.shared().sendEvent(event)
            true
        }

        airship.pushManager.addPushListener { pushMessage, notificationPosted ->
            if (!notificationPosted) {
                val event: Event = PushReceivedEvent(pushMessage)
                EventEmitter.shared().sendEvent(event)
            }
        }

        airship.channel.addChannelListener(object : AirshipChannelListener {
            override fun onChannelCreated(channelId: String) {
                val event: Event = ChannelCreatedEvent(channelId, UAirship.shared().pushManager.pushToken)
                EventEmitter.shared().sendEvent(event)
            }

            override fun onChannelUpdated(channelId: String) {
            }
        })

        airship.pushManager.addPushTokenListener { token ->
            val event = PushTokenReceivedEvent(token)
            EventEmitter.shared().sendEvent(event)
        }

        airship.pushManager.notificationListener = object : NotificationListener {
            override fun onNotificationPosted(notificationInfo: NotificationInfo) {
                val event: Event = PushReceivedEvent(notificationInfo)
                EventEmitter.shared().sendEvent(event)
            }

            override fun onNotificationOpened(notificationInfo: NotificationInfo): Boolean {
                val event: Event = NotificationResponseEvent(notificationInfo, null)
                EventEmitter.shared().sendEvent(event)
                return false
            }

            override fun onNotificationForegroundAction(notificationInfo: NotificationInfo, notificationActionButtonInfo: NotificationActionButtonInfo): Boolean {
                val event: Event = NotificationResponseEvent(notificationInfo, notificationActionButtonInfo)
                EventEmitter.shared().sendEvent(event)
                return false
            }

            override fun onNotificationBackgroundAction(notificationInfo: NotificationInfo, notificationActionButtonInfo: NotificationActionButtonInfo) {
                val event: Event = NotificationResponseEvent(notificationInfo, notificationActionButtonInfo)
                EventEmitter.shared().sendEvent(event)
            }

            override fun onNotificationDismissed(notificationInfo: NotificationInfo) {}
        }

        // Register a listener for inbox update event
        MessageCenter.shared().inbox.addListener {
            val event: Event = MessageCenterUpdatedEvent(MessageCenter.shared().inbox.unreadCount, MessageCenter.shared().inbox.count)
            EventEmitter.shared().sendEvent(event)
        }

        val preferences = ReactAirshipPreferences.shared(context)
        MessageCenter.shared().setOnShowMessageCenterListener { messageId ->
            if (preferences.isAutoLaunchMessageCenterEnabled) {
                false
            } else {
                sendShowInboxEvent(messageId)
                true
            }
        }

        PreferenceCenter.shared().openListener =
            PreferenceCenter.OnOpenListener { preferenceCenterId: String ->
                if (preferences.isAutoLaunchPreferenceCenterEnabled(preferenceCenterId)) {
                    return@OnOpenListener false
                } else {
                    val event: Event = DisplayPreferenceCenterEvent(preferenceCenterId)
                    EventEmitter.shared().sendEvent(event)
                    return@OnOpenListener true
                }
            }

        // Set our custom notification provider
        val notificationProvider = ReactNotificationProvider(context, airship.airshipConfigOptions)
        airship.pushManager.notificationProvider = notificationProvider

        airship.analytics.registerSDKExtension(Analytics.EXTENSION_REACT_NATIVE, BuildConfig.MODULE_VERSION)

        loadCustomNotificationChannels(context, airship)
        loadCustomNotificationButtonGroups(context, airship)

        val extender = createExtender(context)
        extender?.onAirshipReady(context, airship)
    }

    private fun loadCustomNotificationChannels(context: Context, airship: UAirship) {
        val packageName = UAirship.getPackageName()
        @XmlRes val resId = context.resources.getIdentifier("ua_custom_notification_channels", "xml", packageName)

        if (resId != 0) {
            PluginLogger.debug("Loading custom notification channels")
            airship.pushManager.notificationChannelRegistry.createNotificationChannels(resId)
        }
    }

    private fun loadCustomNotificationButtonGroups(context: Context, airship: UAirship) {
        val packageName = UAirship.getPackageName()
        @XmlRes val resId = context.resources.getIdentifier("ua_custom_notification_buttons", "xml", packageName)

        if (resId != 0) {
            PluginLogger.debug("Loading custom notification button groups")
            airship.pushManager.addNotificationActionButtonGroups(context, resId)
        }
    }

    override fun isReady(context: Context): Boolean {
        configOptions = loadConfig(context)

        return try {
            configOptions!!.validate()
            true
        } catch (e: Exception) {
            false
        }
    }

    override fun createAirshipConfigOptions(context: Context): AirshipConfigOptions? {
        return configOptions
    }

    @SuppressLint("RestrictedApi")
    private fun loadConfig(context: Context): AirshipConfigOptions {
        Logger.setLogLevel(Log.ASSERT)
        val builder = AirshipConfigOptions.newBuilder()
            .applyDefaultProperties(context)
            .setRequireInitialRemoteConfigEnabled(true)

        Logger.setLogLevel(Log.ERROR)

        val config: JsonMap? = ReactAirshipPreferences.shared(context).airshipConfig
        if (config == null || config.isEmpty) {
            return builder.build()
        }

        val developmentEnvironment = config.opt("development").map
        val productionEnvironment = config.opt("production").map
        val defaultEnvironment = config.opt("default").map

        developmentEnvironment?.let {
            builder.setDevelopmentAppKey(it.opt("appKey").string)
                .setDevelopmentAppSecret(it.opt("appSecret").string)
            val logLevel = it.opt("logLevel").string
            logLevel?.let { logLvl ->
                builder.setLogLevel(convertLogLevel(logLvl, Log.DEBUG))
            }
        }

        productionEnvironment?.let {
            builder.setProductionAppKey(it.opt("appKey").string)
                .setProductionAppSecret(it.opt("appSecret").string)
            val logLevel = it.opt("logLevel").string
            logLevel?.let { logLvl ->
                builder.setProductionLogLevel(convertLogLevel(logLvl, Log.ERROR))
            }
        }

        defaultEnvironment?.let {
            builder.setAppKey(it.opt("appKey").string)
                .setAppSecret(it.opt("appSecret").string)
            val logLevel = it.opt("logLevel").string
            logLevel?.let { logLvl ->
                builder.setLogLevel(convertLogLevel(logLvl, Log.ERROR))
            }
        }

        val site = config.opt("site").string
        site?.let {
            try {
                builder.setSite(parseSite(it))
            } catch (e: Exception) {
                PluginLogger.error("Invalid site $it", e)
            }
        }

        if (config.containsKey("inProduction")) {
            builder.setInProduction(config.opt("inProduction").getBoolean(false))
        }

        if (config.containsKey("isChannelCreationDelayEnabled")) {
            builder.setChannelCreationDelayEnabled(
                config.opt("isChannelCreationDelayEnabled").getBoolean(false)
            )
        }

        val initialConfigUrl = config.opt("initialConfigUrl").string
        initialConfigUrl?.let {
            try {
                builder.setInitialConfigUrl(initialConfigUrl)
            } catch (e: Exception) {
                PluginLogger.error("Invalid initialConfigUrl $it", e)
            }
        }

        val urlAllowList = parseArray(config.opt("urlAllowList"))
        urlAllowList?.let {
            builder.setUrlAllowList(it)
        }

        val urlAllowListScopeJavaScriptInterface = parseArray(config.opt("urlAllowListScopeJavaScriptInterface"))
        urlAllowListScopeJavaScriptInterface?.let {
            builder.setUrlAllowListScopeJavaScriptInterface(it)
        }

        val urlAllowListScopeOpenUrl = parseArray(config.opt("urlAllowListScopeOpenUrl"))
        urlAllowListScopeOpenUrl?.let {
            builder.setUrlAllowListScopeOpenUrl(it)
        }

        val chat = config.opt("chat").map
        chat?.let {
            builder.setChatSocketUrl(it.opt("webSocketUrl").optString())
                .setChatUrl(it.opt("url").optString())
        }

        val android = config.opt("android").map
        android?.let {
            if (it.containsKey("appStoreUri")) {
                builder.setAppStoreUri(Uri.parse(it.opt("appStoreUri").optString()))
            }

            if (it.containsKey("fcmFirebaseAppName")) {
                builder.setFcmFirebaseAppName(it.opt("fcmFirebaseAppName").optString())
            }

            if (it.containsKey("notificationConfig")) {
                applyNotificationConfig(context, it.opt("notificationConfig").optMap(), builder)
            }
        }

        val enabledFeatures = config.opt("enabledFeatures").list
        try {
            enabledFeatures?.let {
                builder.setEnabledFeatures(parseFeatures(it))
            }
        } catch (e: Exception) {
            PluginLogger.error("Invalid enabled features: $enabledFeatures")
        }
        return builder.build()
    }

    private fun applyNotificationConfig(context: Context, notificationConfig: JsonMap, builder: AirshipConfigOptions.Builder) {
        val icon = notificationConfig.opt(UrbanAirshipReactModule.NOTIFICATION_ICON_KEY).string
        icon?.let {
            val resourceId = getNamedResource(context, it, "drawable")
            builder.setNotificationIcon(resourceId)
        }

        val largeIcon = notificationConfig.opt(UrbanAirshipReactModule.NOTIFICATION_LARGE_ICON_KEY).string
        largeIcon?.let {
            val resourceId = getNamedResource(context, it, "drawable")
            builder.setNotificationLargeIcon(resourceId)
        }

        val accentColor = notificationConfig.opt(UrbanAirshipReactModule.ACCENT_COLOR_KEY).string
        accentColor?.let {
            builder.setNotificationAccentColor(getHexColor(it, 0))
        }

        val channelId = notificationConfig.opt(UrbanAirshipReactModule.DEFAULT_CHANNEL_ID_KEY).string
        channelId?.let {
            builder.setNotificationChannel(it)
        }
    }

    @PrivacyManager.Feature
    private fun parseFeatures(jsonList: JsonList): Int {
        var result = PrivacyManager.FEATURE_NONE
        for (value in jsonList) {
            result = result or parseFeature(value.optString())
        }
        return result
    }

    @AirshipConfigOptions.Site
    private fun parseSite(value: String): String {
        when (value) {
            "eu" -> return AirshipConfigOptions.SITE_EU
            "us" -> return AirshipConfigOptions.SITE_US
        }
        throw IllegalArgumentException("Invalid site: $value")
    }

    companion object {
        const val EXTENDER_MANIFEST_KEY = "com.urbanairship.reactnative.AIRSHIP_EXTENDER"

        private fun sendShowInboxEvent(messageId: String?) {
            val event: Event = DisplayMessageCenterEvent(messageId)
            EventEmitter.shared().sendEvent(event)
        }

        private fun createExtender(context: Context): AirshipExtender? {
            val ai: ApplicationInfo
            try {
                ai = context.packageManager.getApplicationInfo(context.packageName, PackageManager.GET_META_DATA)

                if (ai.metaData == null) {
                    return null
                }
            } catch (e: PackageManager.NameNotFoundException) {
                return null
            }

            val classname = ai.metaData.getString(EXTENDER_MANIFEST_KEY) ?: return null

            try {
                val extenderClass = Class.forName(classname)
                return extenderClass.newInstance() as AirshipExtender
            } catch (e: Exception) {
                PluginLogger.error(e, "Unable to create extender: $classname")
            }
            return null
        }

        private fun convertLogLevel(logLevel: String, defaultValue: Int): Int {
            when (logLevel) {
                "verbose" -> return Log.VERBOSE
                "debug" -> return Log.DEBUG
                "info" -> return Log.INFO
                "warning" -> return Log.WARN
                "error" -> return Log.ERROR
                "none" -> return Log.ASSERT
            }
            return defaultValue
        }

        private fun parseArray(value: JsonValue?): Array<String?>? {
            if (value == null || !value.isJsonList) {
                return null
            }

            val result = arrayOfNulls<String>(value.optList().size())
            for (i in 0 until value.optList().size()) {
                val string = value.optList()[i].string
                if (string == null) {
                    PluginLogger.error("Invalid string array: $value")
                    return null
                }
                result[i] = string
            }
            return result
        }
    }
}